dockertest/subtest.py

```diff
@@ -19,16 +19,45 @@
 from ConfigParser import Error
 from autotest.client.shared.error import TestError, TestNAError
 from autotest.client.shared.version import get_version
-from autotest.client import test
+from autotest.client import test, utils
 import version
 import config
 import subtestbase
+from docker_daemon import which_docker
 from xceptions import DockerTestFail
 from xceptions import DockerTestNAError
 from xceptions import DockerTestError
 from xceptions import DockerSubSubtestNAError
 
 
+def docker_rpm():
+    """
+    Returns the full NVRA of the currently-installed docker or docker-latest
+    """
+    return utils.run("rpm -q %s" % which_docker()).stdout.strip()
+
+
+def known_failures():
+    """
+    Returns a dict containing known test failures. Primary key is
+    subtest name (e.g. docker_cli/foo/bar), value is another dict
+    whose key is docker NVRA (e.g. docker-1.12.5-8.el7.x86_64),
+    value of that is a string description of the problem (e.g.
+    a bz number and comment).
+    """
+    known = {}
+    import csv
+    path_known = os.path.join(config.CONFIGCUSTOMS, 'known_failures.csv')
+    with open(path_known, 'rb') as csv_fh:
+        csv_reader = csv.reader(csv_fh)
+        for row in csv_reader:
+            if row[1] not in known:
+                known[row[1]] = {}
```

Comment:
> I think you need an else: raise(..) here. Without one, this will silently ignore duplicate entries where the docker NVRA happens to clash.

I'm a little confused; I'll try a brief run-through of how the code works, and let's see if we can get on common ground.

The known_failures.csv file is laid out with NVR first. That's because I, as maintainer of this file, want to see docker versions first. It's the most important for me, it lets me quickly see if there are records for the version(s) of docker that I'm testing in a given moment; it lets me add them quickly in one go; it gives me an at-a-glance insight into how buggy a given version may be. The subsubtest is second, then a reason string.

For purposes of the code, the most important field is the subsubtest. That lets the postprocessing code instantly check: "are there known problems with this subsubtest? No? Good, return." Then, if there are known problems, check NVR.

In diagram form:

    docker-1.12.5-*,docker_cli/negativeusage/iv4,"bz1393572: expect fix in docker-1.13?"

becomes:

  known['docker_cli/negativeusage/iv4'] = { 'docker-1.12.5-*': "bzetcetc" }

Does that help? If your comment still applies after this, could you rephrase it or add clarification? Thanks.